src/devices/Charlieplex/CharlieplexSegment.cs

```diff
@@ -0,0 +1,244 @@
+
+                    _controller.Write(node.Anode, node.Value);
+                    // It is necessary to sleep for the LED to be seen with full brightness
+                    // It may be possible to sleep less than 1ms -- this API has ms granularity
+                    Thread.Sleep(1);
+                    _controller.Write(node.Anode, 0);
+                    _lastNode.Anode = node.Anode;
+                    _lastNode.Cathode = node.Cathode;
+                }
+            }
+            while (watch.Elapsed < duration);
+            watch.Stop();
+
+        }
+
+        /// <summary>
+        /// Provides the set of Charlie nodes given the set of pins and the count provided.
+        /// If count = 0, then the Charlieplex maximum is used for the pins provided (n^2-n).
+        /// </summary>
+        /// <param name="pins">The pins to use for the segment.</param>
+        /// <param name="nodeCount">The number of nodes to use. Default is the Charlieplex maximum.</param>
+        public static Node[] GetNodes(int[] pins, int nodeCount = 0)
+        {
+            var pinCount = pins.Length;
+
+            if (nodeCount == 0)
+            {
+                var charlieCount = (int)Math.Pow(pinCount, 2) - pinCount;
+                nodeCount = charlieCount;
+            }
+
+            var charliePins = new Node[nodeCount];
+
+            var pin = 0;
+            var pinJump = 1;
+            var firstLeg = false;
+            var resetCount = pinCount - 1;
+            for (int i = 0; i < nodeCount; i++)
+            {
+                if ((pin > 0 && pin % resetCount == 0) || pin + pinJump > resetCount)
+                {
+                    pin = 0;
+                    pinJump++;
+                }
+
+                var charliePin = new Node();
+
+                if (!firstLeg)
+                {
+                    charliePin.Anode = pins[pin];
+                    charliePin.Cathode = pins[pin + pinJump];
+                    firstLeg = true;
+                }
+                else
+                {
+                    charliePin.Anode = pins[pin + pinJump];
+                    charliePin.Cathode = pins[pin];
+                    firstLeg = false;
+                    pin++;
+                }
+
+                charliePins[i] = charliePin;
+            }
+
+            return charliePins;
+        }
+
+        /// <summary>
+        /// Cleanup.
+        /// Failing to dispose this class, especially when callbacks are active, may lead to undefined behavior.
+        /// </summary>
+        public void Dispose()
+        {
+            // this condition only applies to GPIO devices
+            if (_shouldDispose)
+            {
+                _controller?.Dispose();
+                _controller = null;
+            }
+        }
+
+        /// <summary>
+        /// Represents a node in a Charlieplexed circuit.
+        /// https://wikipedia.org/wiki/Charlieplexing
+        /// </summary>
+        public struct Node
```

Comment:
NIT: Should it make sense to extract this struct out of the class instead? I ask specifically since you have a method that is basically returning you an array of these so it is a type that is exposed by your API. No strong feelings about this so I'm fine with what you choose

Reply:
I think one of two options make sense:

- Keep as-is
- Pull out and then rename to `CharlieplexSegmentNode`

This seems very similar to what I did with the PinMapping type for the shift register: https://github.com/dotnet/iot/blob/master/src/devices/Sn74hc595/Sn74hc595.cs#L272

It's the pattern where there is a type that is both tightly bound to another type and that has a very narrow role to play. That seems like it is well suited for a nest type. Is there good API design guidance for that case?

Reply:
I talked to @terrajobst. He said go with option 2 above. I did.